### PR TITLE
Replace fibers with threadpool

### PR DESCRIPTION
**Pull request by markaren**
This PR replaces #671. It uses a threadpool rather than `std::for_each` so that compilers older than gcc9 will work.

Note that slave state is unimplemented in this PR. It will eventually be included when I have figured out how to best make it back in after removing `async_slave`. 
Additionally the `concurrency` test has been commented out. The test is about file locking.

### CODE REVIEW

File: src/cosim/utility/thread_pool.hpp
Comment (davidhjp01) [inline src/cosim/utility/thread_pool.hpp:42]: Thanks for working on this PR. Would it be possible to have an option to run fmus without spawning a separate worker thread? (i.e. running everything in a main thread sequentially). For some use-cases, the communication overhead between the main and the worker threads seem quite significant.

Comment (kyllingstad) [reply src/cosim/utility/thread_pool.hpp:42]: This is possible today, and was in fact the motivation for using fibers in the first place. (Not that fibers are required for sequential execution of FMU functions, but they provide a nice framework for combining sequential execution with asynchronous I/O so one can run both local and remote slaves without the overhead of threads.)

Comment (kyllingstad) [reply src/cosim/utility/thread_pool.hpp:42]: The problem is just that the remote-slave implementation we're using today (proxyfmu) is not built around async I/O, so it requires a worker thread per slave anyway. That is why fibers simply are an extra overhead that comes _in addition to_ thread and I/O today, which is what I suppose motivated @markaren's PR.

As I see it, there are two ways to improve performance in this area: Drop fibers and go all-in for proxyfmu and one-thread-per-slave (as this PR proposes), or keep fibers and implement async communication with slaves (as was the original plan).

Comment (davidhjp01) [reply src/cosim/utility/thread_pool.hpp:42]: Thanks for the explanation @kyllingstad.

I was able to achieve required simulation performance by removing `std::thread` for slaves as well as recursive fiber creation in the libcosim master branch. I was not able to achieve this via `pseudo_async_slave` because it still creates fiber recursively in the fmu interface methods (calling get/setters for variables and do_step).

But now the problem is I cannot not simulate some fmus because they run directly in fibers with limited resources (created by `slave_simulator::impl::do_step`). And I cannot replace/extend `slave_simulator`  as it is always added in `execution::add_slave`.

Comment (markaren) [reply src/cosim/utility/thread_pool.hpp:42]: > Would it be possible to have an option to run fmus without spawning a separate worker thread?
Yes, and this is one of the motivations for this PR. IMO the master algorithm itself should decide how it handles execution. In the case of this implementation it could take `numThreads` as an argument where < 1 means no pool.

proxy-fmu was not a motivation for this PR. My motivation was to simplify the code base and make it run faster. I've found that the fiber solution is slower than a threaded solution with or without proxy-fmu.

**Comment by markaren** [comment]
> It uses a threadpool rather than std::for_each so that compilers older than gcc9 will work

This is an important thing to address if this is something we want to move forward with. Do we have to support old compilers that are not C++17 feature complete?

**Comment by markaren** [comment]
If the target branch gets the ok, we should also make it possible to specify the count in the XML config.

**Comment by markaren** [comment]
Adding the functionality to set the number of worker threads for SSP was trivial, but not so much for the OSP alternative as it does not provide algorithm spesific configuration options. This is related to #404

**Comment by ljamt** [comment]
> Adding the functionality to set the number of worker threads for SSP was trivial, but not so much for the OSP alternative as it does not provide algorithm spesific configuration options. This is related to #404

I see your point, but I don't think this can be solved separately. Should not be a blocker for this PR.

**Comment by ljamt** [comment]
What's the reason for this PR to remain in draft mode? Any remaining issues that must be resolved?

**Comment by markaren** [comment]
> What's the reason for this PR to remain in draft mode? Any remaining issues that must be resolved?

Not other than significant vetting.

**Review by restenb — CHANGES_REQUESTED** [review]
Great job getting a PR out with fibers removed! 

I left some review comments, particularly around the new thread pool functionality. The implementation looks  correct, but suggestions for improvements.

File: src/cosim/utility/thread_pool.hpp
Comment (davidhjp01) [reply src/cosim/utility/thread_pool.hpp:43]: This might causing the simulation (fmu steps) not parallelized since other worker threads are blocked until the mutex gets released.

File: src/cosim/algorithm/fixed_step_algorithm.cpp
Comment (markaren) [reply src/cosim/algorithm/fixed_step_algorithm.cpp:54]: That would by default decrease the performance, which rarely would be what you want? Users that want to use less threads could always do that in configuration.

File: src/cosim/utility/thread_pool.hpp
Comment (markaren) [reply src/cosim/utility/thread_pool.hpp:87]: FunctionType could be a named function. I'm unsure if that makes a difference.

Comment (markaren) [reply src/cosim/utility/thread_pool.hpp:87]: Ahh, I guess the reason for writing it like that is that FunctionType does not necessary have the signature `std::function<void()>`.

File: src/cosim/utility/thread_pool.hpp
Comment (markaren) [reply src/cosim/utility/thread_pool.hpp:87]: This executes on the main thread. So, yes there is a difference.

Comment (markaren) [reply src/cosim/utility/thread_pool.hpp:87]: The alternative would be to put this logic in `fixed_step_algorithm`

File: src/cosim/utility/thread_pool.hpp
Comment (markaren) [reply src/cosim/utility/thread_pool.hpp:46]: No?

File: src/cosim/utility/thread_pool.hpp
Comment (davidhjp01) [reply src/cosim/utility/thread_pool.hpp:43]: I did not create a separate PR but I see the simulation runs faster with the following modification:
```patch
Index: src/cosim/utility/thread_pool.hpp
<+>UTF-8
===================================================================
diff --git a/src/cosim/utility/thread_pool.hpp b/src/cosim/utility/thread_pool.hpp
--- a/src/cosim/utility/thread_pool.hpp	(revision HEAD)
+++ b/src/cosim/utility/thread_pool.hpp	(revision Staged)
@@ -30,23 +30,27 @@
     std::vector<std::thread> threads_;
     std::mutex m_;
     std::condition_variable cv_;
+    std::condition_variable wk_cv_;
+    uint32_t num_tasks;
 
     void worker_thread()
     {
+        std::chrono::seconds wait_duration(1);
         while (!done_) {
-
             std::unique_lock<std::mutex> lck(m_);
+
             if (!work_queue_.empty()) {
-
                 auto task = work_queue_.front();
                 work_queue_.pop();
-                task();
-
                 lck.unlock();
+                task();
+
+                lck.lock();
+                num_tasks--; // This needs to be decremented after the task is actually completed
                 cv_.notify_one();
-
             } else {
                 std::this_thread::yield();
+                wk_cv_.wait_for(lck, wait_duration);
             }
         }
     }
@@ -54,6 +58,7 @@
 public:
     explicit thread_pool(unsigned int thread_count = std::thread::hardware_concurrency())
         : done_(false)
+        , num_tasks(0)
     {
         try {
             for (unsigned i = 0; i < thread_count; ++i) {
@@ -73,7 +78,9 @@
     {
         if (!threads_.empty()) {
             std::unique_lock<std::mutex> lck(m_);
-            while (!work_queue_.empty()) cv_.wait(lck);
+            while (num_tasks > 0) {
+                cv_.wait(lck);
+            }
         }
     }
 
@@ -84,14 +91,16 @@
             f();
         } else {
             std::unique_lock<std::mutex> lck(m_);
+            num_tasks++;
             work_queue_.push(std::function<void()>(f));
+            wk_cv_.notify_one();
         }
     }
 
     ~thread_pool()
     {
         done_ = true;
-
+        wk_cv_.notify_all();
         for (auto& thread : threads_) {
             if (thread.joinable()) {
                 thread.join();

```

File: src/cosim/utility/thread_pool.hpp
Comment (restenb) [reply src/cosim/utility/thread_pool.hpp:46]: The idea being to wake up one of the worker threads immediately after adding to the work queue, and in the same function, rather than waiting for one of the running threads to process the queue. 

This strikes me as a cleaner implementation that also removes the need to loop all the worker threads constantly. They can simply block until something calls `submit()`, then be notified from here.

File: src/cosim/utility/thread_pool.hpp
Comment (restenb) [reply src/cosim/utility/thread_pool.hpp:46]: Adding my suggestions in code as well (may include typos since I just wrote this into the text editor);

Changes to `worker_thread()`:
```
void worker_thread()
{
    while (true)
    {
        {
            std::unique_lock<std::mutex> lck(m_);
            cv_.wait(lck, [=]{return done || !work_queue_.empty()});
    
            if (done && work_queue_.empty()) break; 

            auto task = std::move(work_queue_.front());
            work_queue_.pop();
        }
        task();
    }
}
```

Changes to `submit()`:

```
template<typename FunctionType>
void submit(FunctionType f)
{
    if (threads_.empty()) {
        f();
    } else {
       {
           std::unique_lock<std::mutex> lck(m_);
           work_queue_.emplace(std::move(std::function<void()>(f)));
       }
      cv_.notify_one();
   }
}
```

File: src/cosim/utility/thread_pool.hpp
Comment (davidhjp01) [reply src/cosim/utility/thread_pool.hpp:95]: This is when `workerThreadCount=0` so we want to invoke the function directly instead of pushing it to the queue (i.e. there is no worker thread).

File: src/cosim/algorithm/fixed_step_algorithm.cpp
Comment (restenb) [reply src/cosim/algorithm/fixed_step_algorithm.cpp:54]: I'd assume you want one thread free for the whatever application you're running that embeds libcosim. 

Though the easier solution I think is to remove the reliance on defaults, and just check user input against `std::thread::hardware_concurrency()` to make sure the requested `workerThreadCount `number can actually be spawned.

File: src/cosim/utility/thread_pool.hpp
Comment (restenb) [reply src/cosim/utility/thread_pool.hpp:95]: Yep, but at that point we don't really have a thread pool. Why make it responsible for running work on the main thread also? 

I think this should be moved out of the thread pool implementation.

File: src/cosim/utility/thread_pool.hpp
Comment (restenb) [inline src/cosim/utility/thread_pool.hpp:53]: I don't really agree with this. Can't we just block the thread until work is available through submit()? This looks like it'll wake the thread up once every second, but I don't immediately see a reason to do so.

Comment (restenb) [reply src/cosim/utility/thread_pool.hpp:53]: That sounds like a synchronization issue? 

I think we need to use the mutex in the destructor also; 

```
std::unique_lock<std::mutex> lck(m_);
done_ = true;
cv_worker_.notify_all();
lck.unlock();
```

Comment (davidhjp01) [reply src/cosim/utility/thread_pool.hpp:53]: From what I understand, `notify_all` can be missed by the worker thread if it is issued while the thread is not in the waiting state (i.e. not blocked by `wait()`). And since `notify_all` can be issued at any arbitrary time (e.g. when the program terminates), it is difficult to control. So it would be safer and cleaner for the thread just to 'timeout' periodically with a reasonable interval to check if `done_=true`.

And yes I think we can guard `done_` in the destructor.

**Review by restenb — CHANGES_REQUESTED** [review]

File: src/cosim/utility/thread_pool.hpp
Comment (restenb) [reply src/cosim/utility/thread_pool.hpp:78]: I committed some changes now, mainly to facilitate the following suggestions: 

* If no work is available, immediately block the worker thread via `cv_worker_`. Revive when we're either `done_`, or `work_queue_ `is not empty (don through `notify_one()` in `submit()`). 
* `wait_for_tasks_to_finish` should wait `pending_tasks_ `and the `work_queue_ ` to empty, not the threads (!)
* Increment pending_tasks_ only when the worker thread "begins" (may not make much of a difference, but seems cleaner) 
* Lock the destructor also via the mutex in order to modify `done_ `and call `notify_all()`. Since all modifications of `done_ `are now in a mutex, it can also be changed to a regular `bool`. In addition I removed the `joinable()` query - if all threads are either running or notified, do we need it? 

I left the previous code in comments for easier comparison. Please have a look.

Comment (davidhjp01) [reply src/cosim/utility/thread_pool.hpp]: I tested the updated version with several test runs, and seems the program terminates correctly (without blocking).
Some comments:
1. Wondering if the worker thread can miss `notify_all` when it is issued while the thread is just before acquiring the lock? (`std::unique_lock<std::mutex> lck(m_)`).
2. The lock does not go out of scope at the end of `if` block unlike the comment mentioned.

File: src/cosim/utility/thread_pool.hpp
Comment (markaren) [reply src/cosim/utility/thread_pool.hpp:98]: Ideally, yes. But it simplifies the logic and amount of code in the algorithm alot. The "pool" is not public API, so we could get away with it even though it breaks the prinicple of cohesion.

File: src/cosim/utility/thread_pool.hpp
Comment (restenb) [reply src/cosim/utility/thread_pool.hpp]: According to the `condition_variable `documentation, it's not necessary to keep the lock when calling `notify_one `or `notify_all`, so I've changed that back to manually calling `unlock()` in both cases.

Comment (restenb) [reply src/cosim/utility/thread_pool.hpp]: Further, the documentation details how `condition_variable `should be used in tandem with a `mutex`, both when modifying a shared variable, and waiting on the condition:

https://en.cppreference.com/w/cpp/thread/condition_variable

I'd say the code is correct now based on this, I can't really spot any deviations from these requirements.

Comment (davidhjp01) [reply src/cosim/utility/thread_pool.hpp]: Seems `wait()` with the pred will run the pred before the thread is put into wait https://en.cppreference.com/w/cpp/thread/condition_variable/wait. Since done_ is guarded, I think this implementation is fine :)

File: include/cosim/slave.hpp
Comment (davidhjp01) [inline include/cosim/slave.hpp:248]: I think we should let the client to decide to whether to re-use this variable for improving performance. For this I made #687.

File: src/cosim/utility/thread_pool.hpp
Comment (davidhjp01) [inline src/cosim/utility/thread_pool.hpp:39]: Will it be possible to move this out of while loop so we don't create/destroy lck for each loop iteration?

**Review by davidhjp01 — COMMENTED** [review]
Good work! Just added some comments

File: src/cosim/utility/thread_pool.hpp
Comment (restenb) [reply src/cosim/utility/thread_pool.hpp:39]: No, I don't think so. How do you then reacquire the lock after a task has been completed, if it's outside of this loop? You'd have to let the thread terminate and recreate it, which we don't want?

**Comment by ljamt** [comment]
`utility_concurrency_unittest` is still commented out. Should be included and fixed before merging.

**Comment by ljamt** [comment]
> `utility_concurrency_unittest` is still commented out. Should be included and fixed before merging.

Test included and passing now

**Review by restenb — APPROVED** [review]

**Review by davidhjp01 — APPROVED** [review]

**Comment by ljamt** [comment]
As fibers are removed, I think `cosim::utility::shared_mutex` can be replaced with `std::shared_mutex`. Don't want to add more to this PR, and can push that as a separate PR.

**Comment by ljamt** [comment]
@markaren, if you are ok with the latest changes to `thread_pool.hpp` are we then ready to merge?
@kyllingstad, @eidekrist, share your opinions if you disagree :)

**Comment by markaren** [comment]
How are the observed differences in usage/speed/accuracy on your side? All good?
We can probably set the the default number of threads to `std::thread::hardware_concurrency()-1` in `fixed_step_algorithm` as suggested?

**Comment by restenb** [comment]
> How are the observed differences in usage/speed/accuracy on your side? All good? We can probably set the the default number of threads to `std::thread::hardware_concurrency()-1` in `fixed_step_algorithm` as suggested?

Yes. I've added a `unsigned int max_threads_ = std::thread::hardware_concurrency() - 1` variable to `fixed_step_algorithm`. Though with what is now a blocking only strategy, it may no longer be necessary. I'd like to extend this in the future to include a spinlock as well though. Blocking & resuming threads has a non-negligible overhead cost when done at a higher rate, such as with a very small timestep simulation. 

We seem to be seeing ~15-20% improvements in simulation speed with this PR over the fiber implementation.

File: include/cosim/algorithm/fixed_step_algorithm.hpp
Comment (kyllingstad) [inline include/cosim/algorithm/fixed_step_algorithm.hpp:37]: `optional` feels heavy-handed here. A more "C++-ish" solution would be to use a native integer type and plain zero instead of `nullopt`.

File: src/cosim/utility/concurrency.hpp
Comment (kyllingstad) [inline src/cosim/utility/concurrency.hpp:32]: I don't think we need this class at all anymore, we can just use `std::shared_mutex`.

**Review by kyllingstad — APPROVED** [review]
Great work! I don't have much to add except some stylistic nitpicks here and there. I didn't look at the thread pool implementation in any detail, since it seems to have been thoroughly reviewed by others. Everything else looks good to merge as far as I'm concerned.

File: src/cosim/utility/concurrency.hpp
Comment (kyllingstad) [reply src/cosim/utility/concurrency.hpp:32]: Nevermind, I see that @ljamt already suggested we do this in a separate PR.

Comment (restenb) [reply src/cosim/utility/concurrency.hpp:32]: Yes, we noticed this as well. In addition `cosim::utility::shared_mutex` was only used in `utility_concurrency_unittest.cpp`, with the intention to test that file locking functioned correctly with the custom mutex, which was in turn only necessary because of fibers. 

With the removal of fibers, that test can be removed as well. We wanted to push this as a separate PR to avoid more noise in this one, but if we all agree they can be removed, might as well do it here.

File: include/cosim/orchestration.hpp
Comment (markaren) [reply include/cosim/orchestration.hpp:194]: Probably just a itchy trigger finger from CLions clang suggestions

File: include/cosim/algorithm/fixed_step_algorithm.hpp
Comment (markaren) [reply include/cosim/algorithm/fixed_step_algorithm.hpp:37]: `std::nullopt` means `std::hardware_currency-1`. 0 means 0. `std::optional` provides more information here.

Comment (kyllingstad) [reply include/cosim/algorithm/fixed_step_algorithm.hpp:37]: Ah, so the number refers to the number of _additional_ threads beyond the main thread? Then I completely agree.

**Comment by kyllingstad** [comment]
One more thing: You may want to consider running clang-format on everything before merging. I see there are some includes that are out of alphabetic order after the `async_slave` --> `slave` change, and possibly other things. If it's not fixed now, it's going to show up in someone else's PR later.

**Pull request merged**
